kernel/of/fdt_mem.c

```diff
@@ -0,0 +1,192 @@
+            *size_ptr = (size_t)(resv_start - start);
+            *split_start_ptr = resv_end;
+            *split_size_ptr = (size_t)(end - *split_start_ptr);
+            return true;
+        }
+        else {
+            // Trim from above
+            end = resv_start;
+            *size_ptr = (size_t)(end - start);
+            return false;
+        }
+    }
+
+    // Should be unreachable
+}
+
+// Try to free this region (return the number of bytes successfully freed)
+static size_t
+try_boot_free_region(struct fdt* fdt, void* start, size_t size)
+{
+    // Should be unnecessary (but it's worth checking in case someone messes up)
+    if (size == 0) {
+        return (size_t)0;
+    }
+
+    // Buffers for a second region if we get split into two
+    void* split_start;
+    size_t split_size;
+
+    // Check if this is a region before the end of the kernel (protects the kernel and
+    // firmware which is before it)
+    extern int __kernel_end[];
+    if (start < (void*)__kernel_end) {
+        if ((void*)(start + size) <= (void*)__kernel_end) {
+            // Entirely before the kernel
+            return (size_t)0;
+        }
+        else {
+            // Need to trim before
+            size = (size_t)((start + size) - (void*)__kernel_end);
+            start = (void*)__kernel_end;
+        }
+    }
+    else {
+        // No addresses lower than the end of the kernel (so no overlap)
+    }
+
+    // Handle the reserve entries in the FDT
+    struct fdt_reserve_entry* entry = fdt_reserve_entry_begin(fdt);
+    while (entry != NULL) {
+        void* resv_start = fdt_reserve_entry_address(entry);
+        size_t resv_size = fdt_reserve_entry_size(entry);
+
+        if (trim_or_split_around_reserved(
+                &start, &size, &split_start, &split_size, resv_start, resv_size
+            )) {
+            // We split into 2 regions
+            return try_boot_free_region(fdt, start, size)
+                   + try_boot_free_region(fdt, split_start, split_size);
+        }
+        else {
+            if (size == 0) {
+                // We got fully trimmed
+                return (size_t)0;
+            }
+            else {
+                // We're still one contiguous region (continue)
+            }
+        }
+
+        entry = fdt_next_reserve_entry(fdt, entry);
+    }
+
+    // Make sure we don't free the FDT blob itself
+    void* fdt_resv_start = (void*)fdt;
+    size_t fdt_resv_size = fdt_size(fdt);
+
+    if (trim_or_split_around_reserved(
+            &start, &size, &split_start, &split_size, fdt_resv_start, fdt_resv_size
+        )) {
+        // We split into 2 regions
+        return try_boot_free_region(fdt, start, size)
+               + try_boot_free_region(fdt, split_start, split_size);
+    }
+    else {
+        if (size == 0) {
+            // We got fully trimmed
+            return (size_t)0;
+        }
+        else {
+            // We're still one contiguous region (continue)
+        }
+    }
```

Comment:
```suggestion
    if (size == 0) {
        // We got fully trimmed
        return 0;
    }
    
    // We're still one contiguous region (continue)
```

try to avoid nesting whenever possible, it significantly improves readability

Reply:
I'm good with this I'll make the changes, I was checking exhaustive cases and forgot to go back and remove the elses